src/Microsoft.IdentityModel.Tokens/Validation/Validators.TokenType.cs

```diff
@@ -57,5 +59,117 @@ public static string ValidateTokenType(string type, SecurityToken securityToken,
+                LogHelper.LogVerbose(LogMessages.IDX10255);
+                return new TokenTypeValidationResult(type);
+            }
+
+            if (validationParameters.TypeValidator != null)
+            {
+                return ValidateTokenTypeUsingDelegate(type, securityToken, validationParameters);
+            }
+
+            // Note: don't return an invalid TokenTypeValidationResult for a null or empty token type when a user-defined delegate is set
+            // to allow it to extract the actual token type from a different location (e.g from the claims).
+            if (string.IsNullOrEmpty(type))
+            {
+                return new TokenTypeValidationResult(
+                    type,
+                    ValidationFailureType.TokenTypeValidationFailed,
+                    new ExceptionDetail(
+                        new MessageDetail(
+                            LogMessages.IDX10256,
+                            LogHelper.MarkAsNonPII(nameof(type))),
+                        typeof(SecurityTokenInvalidTypeException),
+                        new StackFrame(true)));
+            }
+
+            if (!validationParameters.ValidTypes.Contains(type, StringComparer.Ordinal))
+            {
+                return new TokenTypeValidationResult(
+                     type,
+                     ValidationFailureType.TokenTypeValidationFailed,
+                     new ExceptionDetail(
+                         new MessageDetail(
+                             LogMessages.IDX10257,
+                             LogHelper.MarkAsNonPII(nameof(type)),
+                             LogHelper.MarkAsNonPII(Utility.SerializeAsSingleCommaDelimitedString(validationParameters.ValidTypes))),
+                         typeof(SecurityTokenInvalidTypeException),
+                         new StackFrame(true)));
+            }
+
+            if (LogHelper.IsEnabled(EventLogLevel.Informational))
+                LogHelper.LogInformation(LogMessages.IDX10258, LogHelper.MarkAsNonPII(type));
+
+            return new TokenTypeValidationResult(type);
+        }
+
+        private static TokenTypeValidationResult ValidateTokenTypeUsingDelegate(string? type, SecurityToken securityToken, TokenValidationParameters validationParameters)
+        {
+            try
+            {
+                var validatedType = validationParameters.TypeValidator(type, securityToken, validationParameters);
+                return new TokenTypeValidationResult(validatedType);
+            }
+#pragma warning disable CA1031 // Do not catch general exception types
+            catch (Exception ex)
+#pragma warning restore CA1031 // Do not catch general exception types
+            {
+                return new TokenTypeValidationResult(
+                    type,
+                    ValidationFailureType.TokenTypeValidationFailed,
+                    new ExceptionDetail(
+                        new MessageDetail(
+                            LogMessages.IDX10259,
```

Comment:
`LogMessages.IDX10259` does not take parameters.
Given we are passing the exception as the `innerException`, I am not sure we need to include more information on this message.

Reply:
I agree it's not necessary but, I think it's worthwhile to have it clear that this was caused by the customers delegate validation instead of our default validation as the innerexception might not contain enough details of the validation overall validation process.  I can remove it if you think it's too much. Do you have a preference?

Reply:
I agree, will leave this as is for now to keep the scope of this PR in line with our current work until we start moving forward with the new approach being discussed now. Thanks for the feedback!